scipy/spatial/ckdtree.pyi

```diff
@@ -1 +1,17 @@
-class cKDTree: ...
+from typing import Union, Optional
+import numpy as np
+
+class cKDTree:
+    def __init__(self,
+                 data: np.ndarray,
+                 leafsize: int = 16,
+                 compact_nodes: bool = True,
+                 copy_data: bool = False,
+                 balanced_tree: bool = True,
+                 boxsize: Optional[Union[np.ndarray, float]] = None) -> None: ...
+
+    def query_pairs(self,
+                    r: float,
+                    p: float = 2.,
+                    eps: float = 0,
+                    output_type: str = 'set') -> Union[set, np.ndarray]: ...
```

Comment:
So with this you can get a more precise return type by using `@overload` and `Literal`. So that would look something like

```python
from typing import overload, Set
from typing_extensions import Literal

@overload
def query_pairs(
    ....
) -> Set: ...

@overload
def query_pairs(
    ....
    output_type: Literal['set']
) -> Set: ...

@overload
def query_pairs(
    ...
    output_type: Literal['ndarray']
) -> np.ndarray: ...
```

The `numpy-stubs` already require `typing_extensions`, so we're fine using them.

Generally unions in return types tend to be a pain (all calling code needs to be prepared to deal with all return types or use ignores, even when you really know what the return type is), so it's best to avoid them if at all possible.

Reply:
Here, I think I disagree--I tend to side with the confusion observed in this mypy issue:
https://github.com/python/mypy/issues/5407

I spent ~20 minutes trying stuff and getting the same issue with `Overloaded function signature 2 will never be matched`.

This seems rather verbose and complex and `mypy` currently passes, so I'd rather not do that for now--maybe cleanup in the future?